cmd/crates/soroban-spec-typescript/fixtures/test_custom_types/src/index.ts

```diff
@@ -231,392 +231,807 @@ function ComplexEnumFromXdr(base64Xdr: string): ComplexEnum {
-        signAndSend,
-        fee,
-        method: 'woid', 
-        
-    };
-    
-    // @ts-ignore Type does exist
-    const response = await invoke(invokeArgs);
-    return ;
-}
-
-export async function val( {signAndSend, fee}: {signAndSend?: boolean, fee?: number} = {signAndSend: false, fee: 100}): Promise<any> {
-    let invokeArgs: InvokeArgs = {
-        signAndSend,
-        fee,
-        method: 'val', 
-        
-    };
-    
-    // @ts-ignore Type does exist
-    const response = await invoke(invokeArgs);
-    return scValStrToJs(response.xdr) as any;
-}
-
-export async function u32_fail_on_even({u32_}: {u32_: u32}, {signAndSend, fee}: {signAndSend?: boolean, fee?: number} = {signAndSend: false, fee: 100}): Promise<Result<u32>> {
-    let invokeArgs: InvokeArgs = {
-        signAndSend,
-        fee,
-        method: 'u32_fail_on_even', 
-        args: [((i) => xdr.ScVal.scvU32(i))(u32_)], 
-    };
-    
-    try {
-        
-    // @ts-ignore Type does exist
-    const response = await invoke(invokeArgs);
-    return new Ok(scValStrToJs(response.xdr) as u32);
-    } catch (e) {
-        //@ts-ignore
-        let err = getError(e.message);
-        if (err) {
-            return err;
-        } else {
-            throw e;
-        }
-    }
+export async function hello<R extends ResponseTypes = undefined>({hello}: {hello: string}, options: {
+  /**
+   * The fee to pay for the transaction. Default: 100.
+   */
+  fee?: number
+  /**
+   * What type of response to return.
+   *
+   *   - `undefined`, the default, parses the returned XDR as `string`. Runs preflight, checks to see if auth/signing is required, and sends the transaction if so. If there's no error and `secondsToWait` is positive, awaits the finalized transaction.
+   *   - `'simulated'` will only simulate/preflight the transaction, even if it's a change/set method that requires auth/signing. Returns full preflight info.
+   *   - `'full'` return the full RPC response, meaning either 1. the preflight info, if it's a view/read method that doesn't require auth/signing, or 2. the `sendTransaction` response, if there's a problem with sending the transaction or if you set `secondsToWait` to 0, or 3. the `getTransaction` response, if it's a change method with no `sendTransaction` errors and a positive `secondsToWait`.
+   */
+  responseType?: R
+  /**
+   * If the simulation shows that this invocation requires auth/signing, `invoke` will wait `secondsToWait` seconds for the transaction to complete before giving up and returning the incomplete {@link SorobanClient.SorobanRpc.GetTransactionResponse} results (or attempting to parse their probably-missing XDR with `parseResultXdr`, depending on `responseType`). Set this to `0` to skip waiting altogether, which will return you {@link SorobanClient.SorobanRpc.SendTransactionResponse} more quickly, before the transaction has time to be included in the ledger. Default: 10.
+   */
+  secondsToWait?: number
+} = {}) {
```

Comment:
can this `options` member be its own type? so that we can separate out the structure/typedefs/docs from the actual method signature

Reply:
The idea here was that doing that doesn't allow the options to appear with the IDE type hover info.  This way devs don't need to look up a type.  And we figured it's pretty much the same since we are generated it anyway.

Reply:
Hmm true, but it does make the generated code annoying to read (though I guess most people wouldn't do that?). Maybe it's better as a jsdoc on the function itself? i.e. something like

```ts
/**
 * @param {OptionsType} [options]
 * @param {number} [options.fee]  - the fee to pay for the transaction (default: 100). 
 * @param {string} [options.responseType] - the type of response to return:  
 *     - `undefined`, the default, parses the returned XDR as `SimpleEnum`. 
 *       Runs preflight, checks to see if auth/signing is required, and sends the 
 *       transaction if so. If there's no error and `secondsToWait` is positive, 
 *       awaits the finalized transaction.
 *   - `'simulated'` will only simulate/preflight the transaction, even if it's a 
 *      change/set method that requires auth/signing. Returns full preflight info.
 *   - `'full'` return the full RPC response, meaning either 1. the preflight info, 
 *     if it's a view/read method that doesn't require auth/signing, or 
 *     2. the `sendTransaction` response, if there's a problem with sending 
 *     the transaction or if you set `secondsToWait` to 0, or 3. the 
 *     `getTransaction` response, if it's a change method with no `sendTransaction` 
 *     errors and a positive `secondsToWait`
 * @param <and so on...>
 **/
export async function simple<R extends ResponseTypes = undefined>({simple}: {simple: SimpleEnum}, options: { ...
```

or is that **also** incompatible with type-hovers?

Anyway, the code/docs are the important part, less-so where they live. I'm just offering suggestions on readability, but the dev UX is far more crucial, so if this accomplishes that, so be it!

Reply:
In a separate vein: there should still probably be an `enum ResponseType` for that parameter to not rely on the "magic strings" `'full'` etc.

Reply:
I think part of the problem here is that we need to add everything in this `fixtures` folder to our [.gitattributes](https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github) file so that it doesn't show up in diffs 😅 

The idea is as Willem said: maximize typeahead usability. Unfortunately, TypeScript doesn't give us a way to "expand" definitions of a type like `enum ResponseType` or `Options`, so people would just see those when they type functions in their editor:

<img width="618" alt="image" src="https://github.com/stellar/soroban-tools/assets/221614/d3358666-03d1-46e6-8221-dc7365a107b0">

Then they'd have to go to the definition (if they know how) or look it up the docs online.

Reply:
Bah okay fair points, thanks for humoring me. The whole point is giving them usable types to work with, so let's continue to maximize that :+1: